compiler/rustc_lint/src/types/improper_ctypes.rs

```diff
@@ -49,3 +97,1385 @@ fn variant_has_complex_ctor(variant: &ty::VariantDef) -> bool {
+            }
+            r @ _ => r,
+        }
+    }
+    /// If the FfiPhantom variant, turns it into a FfiUnsafe version.
+    /// Otherwise, keep unchanged.
+    fn forbid_phantom(self) -> Self {
+        match self {
+            Self::FfiSafe | Self::FfiUnsafe(..) => self,
+            Self::FfiPhantom(ty) => Self::FfiUnsafe(vec![FfiUnsafeReason {
+                ty,
+                reason: fluent::lint_improper_ctypes_only_phantomdata,
+                help: None,
+                inner: None,
+            }]),
+        }
+    }
+}
+
+impl<'tcx> std::ops::AddAssign<FfiResult<'tcx>> for FfiResult<'tcx> {
+    fn add_assign(&mut self, mut other: Self) {
+        // note: we shouldn't really encounter FfiPhantoms here, they should be dealt with beforehand
+        // still, this function deals with them in a reasonable way, I think
+
+        // this function is awful to look but that's because matching mutable references consumes them (?!)
+        // the function itself imitates the following piece of non-compiling code:
+
+        // match (self, other) {
+        //     (Self::FfiUnsafe(_), _) => {
+        //         // nothing to do
+        //     },
+        //     (_, Self::FfiUnsafe(_)) => {
+        //         *self = other;
+        //     },
+        //     (Self::FfiPhantom(ref ty1),Self::FfiPhantom(ty2)) => {
+        //         println!("whoops, both FfiPhantom: self({:?}) += other({:?})", ty1, ty2);
+        //     },
+        //     (Self::FfiSafe,Self::FfiPhantom(_)) => {
+        //         *self = other;
+        //     },
+        //     (_, Self::FfiSafe) => {
+        //         // nothing to do
+        //     },
+        // }
+
+        let s_disc = std::mem::discriminant(self);
+        let o_disc = std::mem::discriminant(&other);
+        if s_disc == o_disc {
+            match (self, &mut other) {
+                (Self::FfiUnsafe(ref mut s_inner), Self::FfiUnsafe(ref mut o_inner)) => {
+                    s_inner.append(o_inner);
+                }
+                (Self::FfiPhantom(ref ty1), Self::FfiPhantom(ty2)) => {
+                    debug!("whoops: both FfiPhantom, self({:?}) += other({:?})", ty1, ty2);
+                }
+                (Self::FfiSafe, Self::FfiSafe) => {}
+                _ => unreachable!(),
+            }
+        } else {
+            if let Self::FfiUnsafe(_) = self {
+                return;
+            }
+            match other {
+                Self::FfiUnsafe(o_inner) => {
+                    // self is Safe or Phantom: Unsafe wins
+                    *self = Self::FfiUnsafe(o_inner);
+                }
+                Self::FfiSafe => {
+                    // self is always "wins"
+                    return;
+                }
+                Self::FfiPhantom(o_inner) => {
+                    // self is Safe: Phantom wins
+                    *self = Self::FfiPhantom(o_inner);
+                }
+            }
+        }
```

Comment:
The following compiles:
```suggestion
        match (self, other) {
            (Self::FfiUnsafe(_), _) => {
                // nothing to do
            }
            (myself, other @ Self::FfiUnsafe(_)) => {
                *myself = other;
            }
            (Self::FfiPhantom(ref ty1), Self::FfiPhantom(ty2)) => {
                println!("whoops, both FfiPhantom: self({:?}) += other({:?})", ty1, ty2);
            }
            (myself @ Self::FfiSafe, other @ Self::FfiPhantom(_)) => {
                *myself = other;
            }
            (_, Self::FfiSafe) => {
                // nothing to do
            }
        }
```